docs/guides/cross-referencing-with-sphinx.rst

```diff
@@ -0,0 +1,278 @@
+
+.. _target to paragraph:
+
+An easy way is just to use the final link of the page/section.
+This works, but it has some disadvantages:
+
+- Links can change, so they are hard to maintain.
+- Links can be verbose, not easy to know the page/section they are linking to.
+- Not easy way to link to specific sections like paragraphs, figures, or code blocks.
+- It only works for the html version of your documentation.
+
+ReStructuredText has a built-in way to linking to elements,
+and Sphinx extends this to make it even more powerful!
+Some advantages of using reStructuredText's references:
+
+- Use a name instead of the final link.
+- Portable between formats: html, PDF, ePub.
+- Sphinx will warn you of invalid references.
+- Cross reference to more than just pages and sections.
+
+.. contents:: Table of contents
+   :local:
+   :backlinks: none
+   :depth: 3
+
+.. _My target:
+
+Explicit targets
+----------------
+
+If you are not familiar with reStructuredText,
+check :doc:`sphinx:usage/restructuredtext/basics` for a quick introduction.
+
+In reStructuredText we have targets and references,
+where the target is the place where a reference links to.
+
+We can reference to any part of a document by adding a target
+in the line before the element we want to link to.
+For example, one way of creating a target to a section is:
+
+.. code-block:: rst
+
+   .. _My target:
+
+   Explicit targets
+   ----------------
+
+Then we can reference the section using ```My target`_``,
+that will be rendered as `My target`_.
+
+Another example, a target to a paragraph:
+
+.. code-block:: rst
+
+   .. _target to paragraph:
+
+   An easy way is just to use the final link of the page/section.
+   This works, but it has some disadvantages:
+
+Then we can reference it using ```target to paragraph`_``,
+that will be rendered as: `target to paragraph`_.
+
+The reference displays the name of the target by default,
+but we can change that with any text
+```my custom text <target to paragraph>`_``,
+that will be rendered as: `my custom text <target to paragraph>`_.
+
+We can also reference to a place *inside* an element,
+this can be done with an _`inline target`.
+For example, a target inside a paragraph:
+
+.. code-block:: rst
+
+   We can also reference to a place *inside* an element,
+   this can be done with an _`inline target`.
+
+Then we can reference it using ```inline target`_``,
+that will be rendered as: `inline target`_.
+
+Implicit targets
+----------------
+
+When we create a section,
+reStructuredText will create a target with the title as the name.
+For example, to reference to the previous section we can use ```Explicit targets`_``,
+that will be rendered as: `Explicit targets`_.
+
+.. note::
+
+   `Footnotes <https://docutils.sourceforge.io/docs/user/rst/quickref.html#footnotes>`_ and
+   `citations <https://docutils.sourceforge.io/docs/user/rst/quickref.html#citations>`_
+   also create implicit targets.
+
+Cross-referencing using roles
```

Comment:
I think it may be worth breaking this into two different pages - one that covers implicit and explicit targets (because they are only on the same page) and one that covers cross-references with roles etc. This is because the page is titled "cross-referencing..." but we don't get to actual cross referencing until about halfway through the content.

Reply:
I wanted to explain some common concepts before using them to cross-reference, not strong opinion though

Reply:
Yeah I agree those concepts are helpful to understand the subsequent material...I just found myself confused that the page is called "cross referencing" but I didn't see any cross-referencing material at first. Maybe you could put the first section underneath a `getting started:` or `background:` header?

Reply:
That's a good idea!